### DISCUSSION TITLE
need help to dockerize

### DISCUSSION THREAD

**Discussion by eleach** [timlrx/tailwind-nextjs-starter-blog#185] [Q&A] [answered]
Hello Timlrx,

Thanks for making this blog available. It is a great source for study.

Would you have any advice for how to dockerize it? I started with the Dockerfile template from nextjs:

   https://nextjs.org/docs/deployment

It builds an image of your app but it does not run. When trying to docker run the image I get the error:

    Error: > Couldn't find a `pages` directory. Please create one under the project root

I think this has something to do with your webpack publicPath. (I wonder why you use this instead  of basePath.) I tried changing publicPath in your config file and the copy source directory in the Dockerfile but could not get it working. If you have any advice on how to dockerize the app I would appreciate it. Thanks again.

**Comment by timlrx** [comment]
Can you post your Dockerfile?

**Reply by eleach** [reply]
Yes. The Dockerfile below is almost identical to the nextjs template.

"docker build . -t blog" builds fine.

"docker run -p 3000:3000 blog" gives the following error:

 UnhandledPromiseRejectionWarning: Error: > Couldn't find a `pages` directory. Please create one under the project root

My guess is something (the build directory and/or your config file) is not getting copied to the right place from the builder image to the final.

-------------

Dockerfile ->


# Install dependencies only when needed
FROM node:14-alpine AS deps

# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
RUN apk add --no-cache libc6-compat
WORKDIR /app
COPY package.json yarn.lock ./
## RUN yarn install --frozen-lockfile
RUN yarn config set strict-ssl false -g && yarn install --frozen-lockfile

# Rebuild the source code only when needed

FROM node:14-alpine AS builder
WORKDIR /app
COPY . .
COPY --from=deps /app/node_modules ./node_modules
RUN yarn build

# Production image, copy all the files and run next
FROM node:14-alpine AS runner
WORKDIR /app

ENV NODE_ENV production

RUN addgroup -g 1001 -S nodejs
RUN adduser -S nextjs -u 1001

# You only need to copy next.config.js if you are NOT using the default configuration

COPY --from=builder /app/next.config.js ./

COPY --from=builder /app/public ./public

COPY --from=builder --chown=nextjs:nodejs /app/.next ./.next
COPY --from=builder /app/node_modules ./node_modules
COPY --from=builder /app/package.json ./package.json

USER nextjs

EXPOSE 3000

# Next.js collects completely anonymous telemetry data about general usage.
# Learn more here: https://nextjs.org/telemetry
# Uncomment the following line in case you want to disable telemetry.
ENV NEXT_TELEMETRY_DISABLED 1

CMD ["yarn", "start"]

**Reply by timlrx** [reply]
1) Please paste it properly in a code block!
2) My repository uses npm, if you are using yarn I hope you `yarn install` to generate the lock file.
3) You probably want to serve the build files, not serve a dev server. If you check my npm scripts that would correspond with `yarn serve`

**Comment by timlrx** [comment, marked as answer]
Here's my working Dockerfile

```
# Install dependencies only when needed
FROM node:14-alpine AS deps
# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
RUN apk add --no-cache libc6-compat
WORKDIR /app
COPY package.json package-lock.json ./
RUN npm install

# Rebuild the source code only when needed
FROM node:14-alpine AS builder
WORKDIR /app
COPY . .
COPY --from=deps /app/node_modules ./node_modules
RUN npm run build

# Production image, copy all the files and run next
FROM node:14-alpine AS runner
WORKDIR /app

ENV NODE_ENV production

RUN addgroup -g 1001 -S nodejs
RUN adduser -S nextjs -u 1001

# You only need to copy next.config.js if you are NOT using the default configuration
COPY --from=builder /app/next.config.js ./
COPY --from=builder /app/public ./public
COPY --from=builder --chown=nextjs:nodejs /app/.next ./.next
COPY --from=builder /app/node_modules ./node_modules
COPY --from=builder /app/package.json ./package.json

USER nextjs

EXPOSE 3000

# Next.js collects completely anonymous telemetry data about general usage.
# Learn more here: https://nextjs.org/telemetry
# Uncomment the following line in case you want to disable telemetry.
# ENV NEXT_TELEMETRY_DISABLED 1

CMD ["npm", "run", "serve"]
```

**Comment by eleach** [comment]
Got it! I ran into a permissions problem on run. So I added the --chown to the COPY to ./public.
 
I also added: RUN npm config set unsafe-perm true

Which might be asking for trouble.

I also have trouble with docker networking on localhost (I'm on Ubuntu). So I have to docker run with --network=host

Thanks again for your attention. (sorry about the markdown formatting on the dockerfile)

**Reply by timlrx** [reply]
No worries, glad that it works!
